File: src/services/download.ts
import type { DownloadTask } from "../types/download";
import type { MediaFormat } from "../types/media";

/** Tauri event emitted by the engine on every task state change. */
export const DOWNLOAD_TASK_EVENT = "download://task";

export type DownloadTaskListener = (task: DownloadTask) => void;

/**
 * High-level contract between the UI and the download engine.
 *
 * The UI never touches the low-level download implementation: it creates a
 * task, selects one of the formats the authorized metadata fetch offered,
 * starts it, and reacts to task updates. The engine owns all streaming,
 * cancellation and file handling.
 */
export interface DownloadService {
  createTask(sourceUrl: string, title: string): Promise<DownloadTask>;
  selectFormat(taskId: string, format: MediaFormat): Promise<DownloadTask>;
  start(taskId: string): Promise<DownloadTask>;
  cancel(taskId: string): Promise<DownloadTask>;
  getTask(taskId: string): Promise<DownloadTask>;
  cleanup(taskId: string): Promise<void>;
  /** Subscribes to task updates. Returns an unsubscribe function. */
  onTaskUpdate(listener: DownloadTaskListener): () => void;
  dispose(): void;
}

File: src-tauri/src/commands/download.rs
use tauri::{AppHandle, State};

use crate::download::{DownloadManager, DownloadTask};
use crate::error::AppResult;
use crate::media::models::MediaFormat;

#[tauri::command]
pub fn create_download_task(
    manager: State<'_, DownloadManager>,
    source_url: String,
    title: String,
) -> AppResult<DownloadTask> {
    manager.create_task(source_url, title)
}

#[tauri::command]
pub fn select_download_format(
    manager: State<'_, DownloadManager>,
    task_id: String,
    format: MediaFormat,
) -> AppResult<DownloadTask> {
    manager.select_format(&task_id, format)
}

#[tauri::command]
pub async fn start_download(
    app: AppHandle,
    manager: State<'_, DownloadManager>,
    task_id: String,
) -> AppResult<DownloadTask> {
    manager.start(&app, &task_id).await
}

#[tauri::command]
pub fn cancel_download(
    app: AppHandle,
    manager: State<'_, DownloadManager>,
    task_id: String,
) -> AppResult<DownloadTask> {
    manager.cancel(&app, &task_id)
}

#[tauri::command]
pub fn get_download_task(
    manager: State<'_, DownloadManager>,
    task_id: String,
) -> AppResult<DownloadTask> {
    manager.get_task(&task_id)
}

#[tauri::command]
pub fn cleanup_download(
    manager: State<'_, DownloadManager>,
    task_id: String,
) -> AppResult<()> {
    manager.cleanup(&task_id)
}